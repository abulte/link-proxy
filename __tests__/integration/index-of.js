const nock = require('nock')

const mongo = require('../../lib/utils/mongo')
const store = require('../../lib/utils/store')
const {upsertLink} = require('../../lib/link')
const analyze = require('../../jobs/check/analyze')

const indexOf = require('../../__test-helpers__/index-of')
const {shapefile} = require('../../__test-helpers__/archive')

const NAME = 'test-index-of'

beforeAll(async () => {
  process.env.S3_BUCKET = NAME

  await store.client.createBucket({Bucket: NAME}).promise()
  await mongo.connect('mongodb://localhost', NAME)
})

afterAll(async () => {
  const objects = await store.client.listObjects({Bucket: NAME}).promise()
  await store.client.deleteObjects({
    Bucket: NAME,
    Delete: {Objects: objects.Contents.map(({Key}) => ({Key}))}
  }).promise()

  await store.client.deleteBucket({Bucket: NAME}).promise()
  await mongo.db.dropDatabase()
  await mongo.disconnect(true)
})

describe(NAME, () => {
  it('should find a shapefile within the zip file of the index-of', async () => {
    nock(`http://${NAME}`).get('/data.zip').reply(200, () => shapefile('data'), {
      'Transfer-Encoding': 'chunked'
    })

    nock(`http://${NAME}`).get('/data2.zip').reply(200, () => shapefile('data2'), {
      'Transfer-Encoding': 'chunked'
    })

    const iof = indexOf(['/data.zip', '/data2.zip'])
    nock(`http://${NAME}`).get('/').reply(200, iof, {
      'Content-Type': 'text/html',
      'Content-Length': iof.length
    })

    const url = `http://${NAME}`
    const {_id} = await upsertLink(url)
    await analyze(_id, url)
  })

  it('should find a shapefile within the zip file of the index-of of the index-of', async () => {
    nock(`http://${NAME}`).get('/sub/data1.zip').reply(200, () => shapefile('data1'), {
      'Transfer-Encoding': 'chunked'
    })

    nock(`http://${NAME}`).get('/sub/data2.zip').reply(200, () => shapefile('data2'), {
      'Transfer-Encoding': 'chunked'
    })

    const iof1 = indexOf(['/sub/data1.zip', '/sub/data2.zip'])
    nock(`http://${NAME}`).get('/sub').reply(200, iof1, {
      'Content-Type': 'text/html',
      'Content-Length': iof1.length
    })

    const iof2 = indexOf('/sub')
    nock(`http://${NAME}`).get('/').reply(200, iof2, {
      'Content-Type': 'text/html',
      'Content-Length': iof2.length
    })

    const url = `http://${NAME}`
    const {_id} = await upsertLink(url)
    await analyze(_id, url)
  })

  it('should find a shapefile listed in the index-of', async () => {
    const files = ['/cool.shp', '/cool.shx', '/cool.dbf', '/cool.prj']
    for (const file of files) {
      nock(`http://${NAME}`).get(file).reply(200, 'foo', {'Transfer-Encoding': 'chunked'})
    }

    const iof = indexOf(files)
    nock(`http://${NAME}`).get('/').reply(200, iof, {
      'Content-Type': 'text/html',
      'Content-Length': iof.length
    })

    const url = `http://${NAME}`
    const {_id} = await upsertLink(url)
    await analyze(_id, url)
  })
})
