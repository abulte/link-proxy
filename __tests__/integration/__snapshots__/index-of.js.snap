// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-index-of should find a shapefile listed in the index-of 1`] = `
Array [
  Object {
    "archive": true,
    "files": Array [
      Object {
        "digest": "sha384-mMEf/f3VQGdrGhN8saIrKnA1DJpEFx1rEYDGvly7LuP3nVMsih3Z7y6OCOdSo7q7",
        "name": "cool.shp",
        "size": 3,
      },
      Object {
        "digest": "sha384-mMEf/f3VQGdrGhN8saIrKnA1DJpEFx1rEYDGvly7LuP3nVMsih3Z7y6OCOdSo7q7",
        "name": "cool.shx",
        "size": 3,
      },
      Object {
        "digest": "sha384-mMEf/f3VQGdrGhN8saIrKnA1DJpEFx1rEYDGvly7LuP3nVMsih3Z7y6OCOdSo7q7",
        "name": "cool.dbf",
        "size": 3,
      },
      Object {
        "digest": "sha384-mMEf/f3VQGdrGhN8saIrKnA1DJpEFx1rEYDGvly7LuP3nVMsih3Z7y6OCOdSo7q7",
        "name": "cool.prj",
        "size": 3,
      },
    ],
    "type": "shp",
  },
]
`;

exports[`test-index-of should find a shapefile within the zip file of the index-of 1`] = `
Array [
  Object {
    "archive": true,
    "files": Array [
      Object {
        "digest": "sha384-+1t5TeJcJ5X34OTBKFd9lDXmWfSP4yyKwNCd6HqLGKChrQodibskoUMz8lbBRABG",
        "name": "house.shp",
        "size": 11,
      },
      Object {
        "digest": "sha384-g/OOiDmdzWdrQElweiK218tYBpzU+Sqpg39YDz/OGKo2bHgTMSfs8PTUFsXaKRtU",
        "name": "house.cpg",
        "size": 11,
      },
      Object {
        "digest": "sha384-N0PxnfQbUbtgRHdg85YR5d23TJyCCw9dkMQ2b7sKrMLplKm2yiK2a+bPBWapI+PA",
        "name": "house.dbf",
        "size": 11,
      },
      Object {
        "digest": "sha384-KjIjW8mDGhYle3IvWHrtMZ3Yng0z3375s8hE8w9gcirOr8Ha1mldECb4q04fg4vc",
        "name": "house.prj",
        "size": 11,
      },
      Object {
        "digest": "sha384-FA1fzHfLn30z4w1QOE8IM3GlS45Xvmodk+MQdzWCkx8ceyOzE3aLBV9Gl7XXZJ4e",
        "name": "house.shx",
        "size": 11,
      },
    ],
    "type": "shp",
  },
  Object {
    "archive": true,
    "files": Array [
      Object {
        "digest": "sha384-+1t5TeJcJ5X34OTBKFd9lDXmWfSP4yyKwNCd6HqLGKChrQodibskoUMz8lbBRABG",
        "name": "hotel.shp",
        "size": 11,
      },
      Object {
        "digest": "sha384-g/OOiDmdzWdrQElweiK218tYBpzU+Sqpg39YDz/OGKo2bHgTMSfs8PTUFsXaKRtU",
        "name": "hotel.cpg",
        "size": 11,
      },
      Object {
        "digest": "sha384-N0PxnfQbUbtgRHdg85YR5d23TJyCCw9dkMQ2b7sKrMLplKm2yiK2a+bPBWapI+PA",
        "name": "hotel.dbf",
        "size": 11,
      },
      Object {
        "digest": "sha384-KjIjW8mDGhYle3IvWHrtMZ3Yng0z3375s8hE8w9gcirOr8Ha1mldECb4q04fg4vc",
        "name": "hotel.prj",
        "size": 11,
      },
      Object {
        "digest": "sha384-FA1fzHfLn30z4w1QOE8IM3GlS45Xvmodk+MQdzWCkx8ceyOzE3aLBV9Gl7XXZJ4e",
        "name": "hotel.shx",
        "size": 11,
      },
    ],
    "type": "shp",
  },
]
`;

exports[`test-index-of should find a shapefile within the zip file of the index-of of the index-of 1`] = `
Array [
  Object {
    "archive": true,
    "files": Array [
      Object {
        "digest": "sha384-+1t5TeJcJ5X34OTBKFd9lDXmWfSP4yyKwNCd6HqLGKChrQodibskoUMz8lbBRABG",
        "name": "data1.shp",
        "size": 11,
      },
      Object {
        "digest": "sha384-g/OOiDmdzWdrQElweiK218tYBpzU+Sqpg39YDz/OGKo2bHgTMSfs8PTUFsXaKRtU",
        "name": "data1.cpg",
        "size": 11,
      },
      Object {
        "digest": "sha384-N0PxnfQbUbtgRHdg85YR5d23TJyCCw9dkMQ2b7sKrMLplKm2yiK2a+bPBWapI+PA",
        "name": "data1.dbf",
        "size": 11,
      },
      Object {
        "digest": "sha384-KjIjW8mDGhYle3IvWHrtMZ3Yng0z3375s8hE8w9gcirOr8Ha1mldECb4q04fg4vc",
        "name": "data1.prj",
        "size": 11,
      },
      Object {
        "digest": "sha384-FA1fzHfLn30z4w1QOE8IM3GlS45Xvmodk+MQdzWCkx8ceyOzE3aLBV9Gl7XXZJ4e",
        "name": "data1.shx",
        "size": 11,
      },
    ],
    "type": "shp",
  },
  Object {
    "archive": true,
    "files": Array [
      Object {
        "digest": "sha384-+1t5TeJcJ5X34OTBKFd9lDXmWfSP4yyKwNCd6HqLGKChrQodibskoUMz8lbBRABG",
        "name": "data2.shp",
        "size": 11,
      },
      Object {
        "digest": "sha384-g/OOiDmdzWdrQElweiK218tYBpzU+Sqpg39YDz/OGKo2bHgTMSfs8PTUFsXaKRtU",
        "name": "data2.cpg",
        "size": 11,
      },
      Object {
        "digest": "sha384-N0PxnfQbUbtgRHdg85YR5d23TJyCCw9dkMQ2b7sKrMLplKm2yiK2a+bPBWapI+PA",
        "name": "data2.dbf",
        "size": 11,
      },
      Object {
        "digest": "sha384-KjIjW8mDGhYle3IvWHrtMZ3Yng0z3375s8hE8w9gcirOr8Ha1mldECb4q04fg4vc",
        "name": "data2.prj",
        "size": 11,
      },
      Object {
        "digest": "sha384-FA1fzHfLn30z4w1QOE8IM3GlS45Xvmodk+MQdzWCkx8ceyOzE3aLBV9Gl7XXZJ4e",
        "name": "data2.shx",
        "size": 11,
      },
    ],
    "type": "shp",
  },
]
`;
